File: cslibrary/110/problems.c
#include "node.h"
#include <assert.h>
#include <stdio.h>

#define MAX_TREE_SIZE 1000
/*
1. build123()
      2
    /   \
   1     3

  a: by calling newNode() three times,
     and using three pointer variables
  b: by calling newNode() three times,
     and using only one pointer variable
  c: by calling insert() three times passing
     it the root pointer to build up the tree
*/
Node build123a() {
  Node root = NewNode(2);
  Node l = NewNode(1);
  Node r = NewNode(3);

  root->left = l;
  root->right = r;

  return root;
}

Node build123b()
{
  Node root = NewNode(2);
  root->left = NewNode(1);
  root->right = NewNode(3);

  return root;
}

Node build123c()
{
  Node root = insert(NULL, 2);
  root = insert(root, 1);
  root = insert(root, 3);
  return root;
}

// 2. size()
int size(Node node){
  if (node == NULL) return 0;
  return size(node->left) + 1 + size(node->right);
}

void testSize()
{
  assert(size(build123a()) == 3);
  assert(size(build123b()) == 3);
  assert(size(build123c()) == 3);
}

// 3. maxDepth()
int max(int a, int b)
{
  if (a > b) return a;
  return b;
}

int maxDepth(Node node)
{
  if (node == NULL) return 0;
  int lDepth, rDepth;
  lDepth = maxDepth(node->left);
  rDepth = maxDepth(node->right);

  return max(lDepth, rDepth) + 1;
}

void testMaxDepth()
{
  assert(maxDepth(NULL) == 0);
  Node n = build123a();
  assert(maxDepth(n) == 2);
  insert(n, 5);
  assert(maxDepth(n) == 3);
}

/*
  4. minValue()
  Given a non-empty binary search tree,
  return the minimum data value found in that tree.
*/

int minValue(Node node){
  assert(node != NULL);
  Node current = node;

  while(current->left != NULL){
    current = current->left;
  }

  return current->data;
}

int minValueRec(Node node)
{
  if (node->left != NULL)
    return minValueRec(node->left);
  return node->data;
}

void testMinValue()
{
  Node n = build123a();
  assert(minValue(n) == 1);
  assert(minValueRec(n) == 1);
  insert(n, 5);
  assert(minValue(n) == 1);
  assert(minValueRec(n) == 1);
  insert(n, -3);
  insert(n, 0);
  assert(minValue(n) == -3);
  assert(minValueRec(n) == -3);
}

/*
  5. printTree()
    The tree...
        4
       / \
      2   5
     / \
    1   3
    Produces the output "1 2 3 4 5"
*/

void printTree(Node node){
  if (node == NULL) return;

  printTree(node->left);
  printf("%d ", node->data);
  printTree(node->right);
}

/*
  6. printPostorder()
    The tree...
        4
       / \
      2   5
     / \
    1   3
    Produces the output "1 3 2 5 4"
*/

void printPostorder(Node node)
{
  if (node == NULL)
    return;

  printPostorder(node->left);
  printPostorder(node->right);
  printf("%d ", node->data);
}

// 7. hasPathSum()
int hasPathSum(Node node, int sum){
  if (node == NULL) return sum == 0;

  int subSum = sum - node->data;
  return hasPathSum(node->left, subSum) || hasPathSum(node->right, subSum);
};

void testHasPathSum(){
  assert(hasPathSum(NULL, 0) == 1);
  assert(hasPathSum(NULL, 1) == 0);
  Node n = build123a();
  assert(hasPathSum(n, 3) == 1);
  assert(hasPathSum(n, 5) == 1);
  assert(hasPathSum(n, 4) == 0);
}

// 8. printPaths()
void printArray(int array[], int len)
{
  int i;
  for(i = 0; i < len; i++){
    printf("%d ", array[i]);
  }
  printf("\n");
}

void printPathsRecur(Node node, int path[], int pathLen)
{
  if (node == NULL)
    return;
  // printf("len: %d, data: %d\n", pathLen, node->data);
  path[pathLen] = node->data;
  pathLen++;

  if (node->left == NULL && node->right == NULL){
    printArray(path, pathLen);
  } else {
    printPathsRecur(node->left, path, pathLen);
    printPathsRecur(node->right, path, pathLen);
  }
}

void printPaths(Node node){
  int path[MAX_TREE_SIZE];
  printPathsRecur(node, path, 0);
}

void testPrints()
{
  Node n = NewNode(4);
  n->left = build123a();
  n->right = NewNode(5);
  printTree(n);
  printf("\n");
  printPostorder(n);
  printf("\n");
  printPaths(n);
}

int main()
{
  testSize();
  testMaxDepth();
  testMinValue();
  testPrints();
  testHasPathSum();
  return 0;
}

File: cslibrary/110/problems.go
package main

import (
	"fmt"
	"os"
)

// Global declarations

func assert(t bool, msg string){
	if (!t){
		fmt.Fprintf(os.Stderr, msg)
		os.Exit(1)
	}
}

type Node struct {
	Data int
	Left *Node
	Right *Node
}

func NewNode(data int) *Node {
	return &Node{
		Data: data,
		Left: nil,
		Right: nil,
	}
}

func (n *Node) Insert(data int) *Node {
	if (n == nil) {
		return NewNode(data)
	}
	if (data <= n.Data) {
		n.Left = n.Left.Insert(data)
	} else {
		n.Right = n.Right.Insert(data)
	}

	return n
}

// Problems
/* 1. build123()
      2
    /   \
   1     3
*/
func build123() *Node {
	var root *Node = nil

	root = root.Insert(2)
	root = root.Insert(1)
	root = root.Insert(3)

	return root
}

// 2. size()
func (n *Node) size() int {
	if (n == nil) {
		return 0
	}
	return n.Left.size() + 1 + n.Right.size()
}

// 3. maxDepth()
func max(a int, b int) int {
	if (a>b) {
		return a
	}
	return b
}

func (n *Node) maxDepth() int {
	if (n == nil){
		return 0
	}
	var lDepth, rDepth int
	lDepth = n.Left.maxDepth()
	rDepth = n.Right.maxDepth()

	return max(lDepth, rDepth) + 1
}

// 4. minValue()
func (n *Node) minValue() *int {
	if (n == nil){
		return nil
	}
	if (n.Left != nil){
		return n.Left.minValue()
	}
	return &n.Data
}

/* 5. printTree()
    The tree...
        4
       / \
      2   5
     / \
    1   3
    Produces the output "1 2 3 4 5"
*/

func (n *Node) printTree() {
	if (n == nil){ return }

	n.Left.printTree()
	fmt.Printf("%d ", n.Data)
	n.Right.printTree()
}

/* 6. printPostorder()
    The tree...
        4
       / \
      2   5
     / \
    1   3
    Produces the output "1 3 2 5 4"
*/

func (n *Node) printPostOrder() {
	if (n == nil){ return }

	n.Left.printPostOrder()
	n.Right.printPostOrder()
	fmt.Printf("%d ", n.Data)
}

func main(){
	var n *Node = nil
	var m *Node = build123()
	// testSize
	assert(n.size() == 0, "Empty node has size 0\n")
	assert(m.size() == 3, "build123 node has size 3\n")
	// testMaxDepth
	assert(n.maxDepth() == 0, "Empty node has maxDepth 0\n")
	assert(m.maxDepth() == 2, "build123 node has maxDepth 2\n")
	// testMinValue
	assert(n.minValue() == nil, "Empty node has minValue nil\n")
	assert(*m.minValue() == 1, "build123 node has minValue 1\n")
	// testPrintTree
	t := &Node{
		Data: 4,
		Left: m,
		Right: NewNode(5),
	}
	fmt.Println("Test printTree:")
	n.printTree()
	fmt.Println()
	m.printTree()
	fmt.Println()
	t.printTree()
	fmt.Println()
	// testPrintPostOrder
	fmt.Println("Test printPostOrder:")
	n.printPostOrder()
	fmt.Println()
	m.printPostOrder()
	fmt.Println()
	t.printPostOrder()
	fmt.Println()
}